File: Binary_Tree/114.Flatten_Binary_Tree_to_Linked_List/Python/114.Flatten_Binary_Tree_to_Linked_List.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: None Do not return anything, modify root in-place instead.
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#

        ############
        #Initialize
        ##### Result array #####
        res_arry = []

        ####################################
        #Recursion-based DFS loop traversal
        self.dfsTraversal(root, res_arry) #Recursion function call (i.e. while root node)

        len_res_arry = len(res_arry) #Length of result array
        len_res_arry -= 1

        for res_node_idx in range(len_res_arry):
            (res_arry[res_node_idx]).right = res_arry[(res_node_idx + 1)] #Keep updating/overwriting


    def dfsTraversal(self, root, res_arry):
        """
        :type root: TreeNode
        :type res_arry: List[TreeNode]
        :rtype: None, void
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#
        if (not root): #Issue/Boundary-case handling
            return 

        root_left, root_right = (root.left), (root.right) #Root left, right half node

        (root.left) = None #Reset
        (root.right) = None #Reset

        res_arry.append(root) #Update/Record

        self.dfsTraversal(root_left, res_arry) #Recursion function call (i.e. root left-half node)
        self.dfsTraversal(root_right, res_arry) #Recursion function call (i.e. root right-half node)

File: Binary_Tree/114.Flatten_Binary_Tree_to_Linked_List/C++/114.Flatten_Binary_Tree_to_Linked_List.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void dfsTraversal(TreeNode * root, vector<TreeNode * > & res_vect)
    {
        //=========================================//
        // Input type:                             //
        //   - TreeNode * (root tree-node)         //
        //   - vector<TreeNode * > (result vector) //
        // Return type:                            //
        //   - void (i.e. no return)               //
        //=========================================//
        if (! root) { return; } //Issue, Boundary-case handling

        /*Whole process, flow*/
        TreeNode * root_left = (root->left), * root_right = (root->right); //Root left, right-half node

        (root->left) = NULL; //Update, Overwrite
        (root->right) = NULL; //Update, Overwrite

        res_vect.emplace_back(root); //Update, Record

        dfsTraversal(root_left, res_vect); //Recursion function call (i.e. root left-half node)
        dfsTraversal(root_right, res_vect); //Recursion function call (i.e. root right-half node)
    }


    void flatten(TreeNode * root) 
    {
        //=================================//
        // Input type:                     //
        //   - TreeNode * (root tree-node) //
        // Return type:                    //
        //   - void (i.e. no return)       //
        //=================================//

        /*Initialize*/
        ///// Length of result vector /////
        int len_res_vect;

        ///// Result vector /////+
        vector<TreeNode * > res_vect;


        /*Recursion-based DFS traversal method*/
        dfsTraversal(root, res_vect); //Recursion function call (i.e. whole node)

        len_res_vect = (res_vect.size()); //Length of result vector
        (len_res_vect--); //Update, Reduce

        for (int res_vect_idx = 0; (res_vect_idx < len_res_vect); (res_vect_idx++)) //Whole
        {
            ((res_vect[res_vect_idx])->right) = res_vect[(res_vect_idx + 1)]; //Keep updating, overwriting

        } //Whole
    }
};
